pil-analyzer/tests/types.rs

```diff
@@ -614,3 +730,50 @@ fn new_fixed_column_wrong_type() {
     "#;
     type_check(input, &[]);
 }
+
+#[test]
+fn trait_with_user_defined_enum() {
+    let input = "
+    enum Bool { True, False }
+    
+    trait Not<T> {
+        not: T -> T,
+    }
+    
+    impl<T> Not<Bool> {
+        not: |b| match b {
+            Bool::True => Bool::False,
+            Bool::False => Bool::True,
+        },
+    }
+    let b: Bool = Not::not(Bool::True);
+    ";
+    type_check(input, &[("b", "", "Bool")]);
+}
+
+#[test]
+fn trait_with_user_defined_enum2() {
+    let input = "
+    enum V1 { A }
+    enum V2 { B }
+
+    trait Convert<T, U> {
+        convert: T -> U,
+    }
+    impl<T, U> Convert<V1, V2> {
+        convert: |x| match x {
+            V1::A => V2::B,
+        },
+    }
+
+    impl<T, U> Convert<V2, V1> {
+        convert: |x| match x {
+            V2::B => V1::A,
+        },
+    }
+
+    let r1: V2 = Convert::convert(V1::A);
```

Comment:
Same here, would be more interesting without explicit type.

Reply:
Why? shouldn't it be possible to derive the type of each let?